openfl/experimental/workflow/component/director/director.py

```diff
@@ -0,0 +1,314 @@
+
+import dill
+
+from openfl.experimental.workflow.component.director.experiment import (
+    Experiment,
+    ExperimentsRegistry,
+)
+from openfl.experimental.workflow.transport.grpc.exceptions import EnvoyNotFoundError
+
+logger = logging.getLogger(__name__)
+
+
+class Director:
+    """Director class for managing experiments and envoys.
+
+    Attributes:
+        tls (bool): A flag indicating if TLS should be used for connections.
+        root_certificate (Optional[Union[Path, str]]): The path to the root certificate
+            for TLS.
+        private_key (Optional[Union[Path, str]]): The path to the private key for TLS.
+        certificate (Optional[Union[Path, str]]): The path to the certificate for TLS.
+        director_config (Optional[Path]): Path to director_config file
+        install_requirements (bool): A flag indicating if the requirements
+            should be installed.
+        _flow_status (Queue): Stores the flow status
+        experiments_registry (ExperimentsRegistry): An object of
+            ExperimentsRegistry to store the experiments.
+        col_exp (dict): A dictionary to store the experiments for
+            collaborators.
+        col_exp_queues (defaultdict): A defaultdict to store the experiment
+            queues for collaborators.
+        _envoy_registry (dict): A dcitionary to store envoy info
+        envoy_health_check_period (int): The period for health check of envoys
+            in seconds.
+        authorized_cols (list): A list of authorized envoys
+    """
+
+    def __init__(
+        self,
+        *,
+        tls: bool = True,
+        root_certificate: Optional[Union[Path, str]] = None,
+        private_key: Optional[Union[Path, str]] = None,
+        certificate: Optional[Union[Path, str]] = None,
+        director_config: Optional[Path] = None,
+        envoy_health_check_period: int = 60,
+        install_requirements: bool = True,
+    ) -> None:
+        """Initialize a Director object.
+
+        Args:
+            tls (bool, optional): A flag indicating if TLS should be used for
+                connections. Defaults to True.
+            root_certificate (Optional[Union[Path, str]]): The path to the
+                root certificate for TLS. Defaults to None.
+            private_key (Optional[Union[Path, str]]): The path to the private
+                key for TLS. Defaults to None.
+            certificate (Optional[Union[Path, str]]): The path to the
+                certificate for TLS. Defaults to None.
+            director_config (Optional[Path]): Path to director_config file
+            envoy_health_check_period (int): The period for health check of envoys
+            in seconds.
+            install_requirements (bool, optional): A flag indicating if the
+                requirements should be installed. Defaults to True.
+        """
+        self.tls = tls
+        self.root_certificate = root_certificate
+        self.private_key = private_key
```

Comment:
I suggest to add an explicit verification that if `tls` is `True`, then a `private_key` object should be present.

Reply:
Same as https://github.com/securefederatedai/openfl/pull/1190#discussion_r1875951703